osbuild/meta.py

```diff
@@ -585,3 +629,50 @@ def get_schema(self, klass, name=None, version="1") -> Schema:
         self._schemata[(klass, name, version)] = schema
 
         return schema
+
+    def list_runners(self, distro: Optional[str] = None) -> List[RunnerInfo]:
+        """List all available runner modules"""
+        if not self._runners:
+            path = os.path.join(self.path, "runners")
+            names = filter(lambda f: os.path.isfile(f"{path}/{f}"),
+                           os.listdir(path))
+            paths = map(lambda n: os.path.join(path, n), names)
+
+            runners_mapped = map(RunnerInfo.from_path, paths)
+            runners_sorted = sorted(runners_mapped, key=lambda r: r.distro)
+            runners_sorted = sorted(runners_sorted, key=lambda r: r.version)
+
+            self._runners = list(runners_sorted)
+
+        runners = self._runners[:]
+        if distro:
+            runners = list(filter(lambda r: r.distro == distro, runners))
+
+        return runners
+
+    def detect_runner(self, name, fallback="org.osbuild.linux") -> RunnerInfo:
+        """Detect the runner for the given name"""
```

Comment:
This comment actually made me think about the current logic a bit more. Currently we use the fallback if there is no runner for the requested distro at all, otherwise we use the version that best matches, even if the version is lower than any runner; i.e. we would return `fedora30` if one requested `fedora29`.

I changed the code that we always opt in and we only support versions newer then a base and fail otherwise. I also added a comment, hopefully should explain things better.